pkg/core/state/notification_event.go

```diff
@@ -79,3 +81,119 @@ func (aer *AppExecResult) DecodeBinary(r *io.BinReader) {
 	}
 	r.ReadArray(&aer.Events)
 }
+
+// notificationEventAux is an auxiliary struct for NotificationEvent JSON marshalling.
+type notificationEventAux struct {
+	ScriptHash util.Uint160    `json:"contract"`
+	Name       string          `json:"eventname"`
+	Item       json.RawMessage `json:"state"`
+}
+
+// MarshalJSON implements implements json.Marshaler interface.
+func (ne *NotificationEvent) MarshalJSON() ([]byte, error) {
+	item, err := stackitem.ToJSONWithTypes(ne.Item)
+	if err != nil {
+		item = []byte(`"error: recursive reference"`)
+	}
+	return json.Marshal(&notificationEventAux{
+		ScriptHash: ne.ScriptHash,
+		Name:       ne.Name,
+		Item:       item,
+	})
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface.
+func (ne *NotificationEvent) UnmarshalJSON(data []byte) error {
+	aux := new(notificationEventAux)
+	if err := json.Unmarshal(data, aux); err != nil {
+		return err
+	}
+	item, err := stackitem.FromJSONWithTypes(aux.Item)
+	if err != nil {
+		return err
+	}
+	if t := item.Type(); t != stackitem.ArrayT {
+		return fmt.Errorf("failed to convert notification event state of type %s to array", t.String())
+	}
+	ne.Item = item.(*stackitem.Array)
+	ne.Name = aux.Name
+	ne.ScriptHash = aux.ScriptHash
+	return nil
+}
+
+// appExecResultAux is an auxiliary struct for JSON marshalling
+type appExecResultAux struct {
+	TxHash      util.Uint256        `json:"txid"`
+	Trigger     string              `json:"trigger"`
+	VMState     string              `json:"vmstate"`
+	GasConsumed int64               `json:"gasconsumed,string"`
+	Stack       json.RawMessage     `json:"stack"`
```

Comment:
`[]json.RawMessage`?

Reply:
We should have the ability to marshal `error: recursive reference` string instead of stack, and this string is not the `[]json.RawMessage`, it's just a `json.RawMessage`. So I don't think that we can use `[]json.RawMessage` here.

Reply:
OK, compatibility...